File: packages/core/src/registry.ts
import { AmbiguousModelError, PolyLLMError, UnknownModelError } from './errors.js'
import type { ModelSpec } from './types.ts'

const normalize = (value: string): string => value.trim().toLowerCase().replace(/[-\s_]/g, '')

export class ModelRegistry {
  private readonly models = new Map<string, ModelSpec>()
  private readonly snapshots = new Map<string, ModelSpec>()
  private readonly aliases = new Map<string, ModelSpec>()
  private readonly normalized = new Map<string, ModelSpec[]>()

  register(spec: ModelSpec): void {
    if (!spec.id || !spec.provider) {
      throw new PolyLLMError('DUPLICATE_MODEL', 'ModelSpec.id 和 ModelSpec.provider 不能为空', { spec })
    }
    const key = `${spec.provider}:${spec.id}`
    if (this.models.has(key)) {
      throw new PolyLLMError('DUPLICATE_MODEL', `模型重复注册: ${key}`, { model: key })
    }
    if (spec.snapshot && this.snapshots.has(`${spec.provider}:${spec.snapshot}`)) {
      throw new PolyLLMError('DUPLICATE_MODEL', `模型快照重复注册: ${spec.provider}:${spec.snapshot}`, {
        model: `${spec.provider}:${spec.snapshot}`,
      })
    }

    this.models.set(key, spec)
    if (spec.snapshot) this.snapshots.set(`${spec.provider}:${spec.snapshot}`, spec)
    for (const alias of spec.aliases ?? []) {
      const aliasKey = `${spec.provider}:${alias}`
      if (this.aliases.has(aliasKey)) {
        throw new PolyLLMError('DUPLICATE_MODEL', `模型别名重复注册: ${aliasKey}`, { alias: aliasKey })
      }
      this.aliases.set(aliasKey, spec)
    }

    const normalizedKey = normalize(`${spec.provider}:${spec.id}`)
    this.normalized.set(normalizedKey, [...(this.normalized.get(normalizedKey) ?? []), spec])
  }

  resolve(input: string): ModelSpec {
    const trimmed = input.trim()
    const separator = trimmed.indexOf(':')
    const explicitProvider = separator > 0 ? trimmed.slice(0, separator) : undefined
    const modelName = separator > 0 ? trimmed.slice(separator + 1) : trimmed
    if (!modelName) throw new UnknownModelError(input, this.listIds())

    if (explicitProvider) {
      const provider = explicitProvider.toLowerCase()
      const snapshot = this.snapshots.get(`${provider}:${modelName}`)
      if (snapshot) return snapshot
      const model = this.models.get(`${provider}:${modelName}`)
      if (model) return model
      const alias = this.aliases.get(`${provider}:${modelName}`)
      if (alias) return alias
      const candidates = this.findNormalized(normalize(`${provider}:${modelName}`))
      if (candidates.length === 1) return candidates[0]
      if (candidates.length > 1) throw new AmbiguousModelError(input, candidates.map(this.modelId))
      throw new UnknownModelError(input, this.listIds())
    }

    const globalSnapshot = this.findByEntry((spec) => spec.snapshot === modelName)
    if (globalSnapshot.length === 1) return globalSnapshot[0]
    if (globalSnapshot.length > 1) throw new AmbiguousModelError(input, globalSnapshot.map(this.modelId))

    const globalModel = this.findByEntry((spec) => spec.id === modelName)
    if (globalModel.length === 1) return globalModel[0]
    if (globalModel.length > 1) throw new AmbiguousModelError(input, globalModel.map(this.modelId))

    const globalAlias = this.findByEntry((spec) => (spec.aliases ?? []).includes(modelName))
    if (globalAlias.length === 1) return globalAlias[0]
    if (globalAlias.length > 1) throw new AmbiguousModelError(input, globalAlias.map(this.modelId))

    const normalized = this.findNormalized(normalize(modelName))
    if (normalized.length === 1) return normalized[0]
    if (normalized.length > 1) throw new AmbiguousModelError(input, normalized.map(this.modelId))
    throw new UnknownModelError(input, this.listIds())
  }

  list(): ModelSpec[] {
    return [...this.models.values()]
  }

  listIds(): string[] {
    return this.list().map(this.modelId)
  }

  private findNormalized(key: string): ModelSpec[] {
    return this.normalized.get(key) ?? []
  }

  private findByEntry(predicate: (spec: ModelSpec) => boolean): ModelSpec[] {
    return this.list().filter(predicate)
  }

  private readonly modelId = (spec: ModelSpec): string => `${spec.provider}:${spec.id}`
}

File: python/packages/core/src/polyllm_core/registry.py
from __future__ import annotations

import re
from typing import List

from .errors import AmbiguousModelError, PolyLLMError, UnknownModelError
from .types import ModelSpec


def _normalize(value: str) -> str:
    return re.sub(r"[-\s_]", "", value.strip().lower())


class ModelRegistry:
    def __init__(self) -> None:
        self._models: dict[str, ModelSpec] = {}
        self._snapshots: dict[str, ModelSpec] = {}
        self._aliases: dict[str, ModelSpec] = {}

    def register(self, spec: ModelSpec) -> None:
        if not spec.id or not spec.provider:
            raise PolyLLMError("DUPLICATE_MODEL", "ModelSpec.id 和 ModelSpec.provider 不能为空", {"spec": spec})

        key = f"{spec.provider}:{spec.id}"
        if key in self._models:
            raise PolyLLMError("DUPLICATE_MODEL", f"模型重复注册: {key}", {"model": key})
        if spec.snapshot:
            snapshot_key = f"{spec.provider}:{spec.snapshot}"
            if snapshot_key in self._snapshots:
                raise PolyLLMError("DUPLICATE_MODEL", f"模型快照重复注册: {snapshot_key}", {"model": snapshot_key})
            self._snapshots[snapshot_key] = spec

        self._models[key] = spec
        for alias in spec.aliases:
            alias_key = f"{spec.provider}:{alias}"
            if alias_key in self._aliases:
                raise PolyLLMError("DUPLICATE_MODEL", f"模型别名重复注册: {alias_key}", {"alias": alias_key})
            self._aliases[alias_key] = spec

    def resolve(self, model: str) -> ModelSpec:
        trimmed = model.strip()
        separator = trimmed.find(":")
        explicit_provider = separator > 0 and trimmed[:separator]
        model_name = trimmed[separator + 1 :] if separator > 0 else trimmed
        if not model_name:
            raise UnknownModelError(model, self.list_ids())

        if explicit_provider:
            provider = explicit_provider.lower()
            candidates = [
                self._snapshots.get(f"{provider}:{model_name}"),
                self._models.get(f"{provider}:{model_name}"),
                self._aliases.get(f"{provider}:{model_name}"),
            ]
            exact = next((item for item in candidates if item is not None), None)
            if exact:
                return exact

            normalized = self._find_normalized(f"{provider}:{model_name}")
            if len(normalized) == 1:
                return normalized[0]
            if len(normalized) > 1:
                raise AmbiguousModelError(model, [self._model_id(item) for item in normalized])
            raise UnknownModelError(model, self.list_ids())

        for predicate in (
            lambda spec: spec.snapshot == model_name,
            lambda spec: spec.id == model_name,
            lambda spec: model_name in spec.aliases,
        ):
            matches = [spec for spec in self._models.values() if predicate(spec)]
            if len(matches) == 1:
                return matches[0]
            if len(matches) > 1:
                raise AmbiguousModelError(model, [self._model_id(spec) for spec in matches])

        normalized = self._find_normalized(model_name)
        if len(normalized) == 1:
            return normalized[0]
        if len(normalized) > 1:
            raise AmbiguousModelError(model, [self._model_id(spec) for spec in normalized])
        raise UnknownModelError(model, self.list_ids())

    def list(self) -> List[ModelSpec]:
        return list(self._models.values())

    def list_ids(self) -> List[str]:
        return [self._model_id(spec) for spec in self._models.values()]

    def _find_normalized(self, value: str) -> List[ModelSpec]:
        normalized = _normalize(value)
        return [spec for spec in self._models.values() if _normalize(f"{spec.provider}:{spec.id}") == normalized]

    @staticmethod
    def _model_id(spec: ModelSpec) -> str:
        return f"{spec.provider}:{spec.id}"
